lib/ProductOpener/Tags.pm

```diff
@@ -1702,6 +1704,188 @@ sub build_tags_taxonomy($$$) {
+
+Properties can be requested with their name (e.g."description") or name + a specific language (e.g. "vegan:en").
+Only properties directly defined for the entry are returned.
+To include inherited properties from parents, prefix the property with "inherited:" (e.g. "inherited:vegan:en").
+
+- include_parents: include entries for all direct parents of the requested tags
+- include_children: include entries for all direct children of the requested tags
+
+=head4 reference to an array of language codes
+
+Languages for which we want to extract names, synonyms, properties.
+
+=cut
+
+sub generate_tags_taxonomy_extract ($$$$) {
+
+	my $tagtype = shift;
+	my $tags_ref = shift;
+	my $options_ref = shift;
+	my $lcs_ref = shift;
+
+	# Return empty hash if the taxonomy does not exist
+	if (not defined $translations_to{$tagtype}) {
+		return {};
+	}
+
+	# For the options include_children or include_parents,
+	# we will need to include data for more tags that requested.
+	# @tags will hold the tags to include
+
+	my @tags = ();
+	my %requested_tags = ();
+	foreach my $tagid (@$tags_ref) {
+		push @tags, $tagid;
+		$requested_tags{$tagid} = 1;
+	}
+
+	my $fields_ref;
+	my @properties = ();
+	if ((defined $options_ref) and (defined $options_ref->{fields})) {
+		$fields_ref = {};
+		foreach my $field (split(/,/, $options_ref->{fields}) ) {
+			# Compute a list of the requested inherited properties,
+			# as we will populate them directly
+			if ($field =~  /^inherited:(.*):(\w\w)$/) {
+				my $prop = $1;
+				my $lc = $2;
+				push @properties, [$prop, $lc];
+			}
+			# Compute a hash of the other requested fields
+			# as we will go through all existing properties
+			else {
+				$fields_ref->{$field} = 1;
+			}
+		}
+	}
+
+	my $taxonomy_ref = {};
+
+	while (my $tagid = shift @tags) {
+
+		if (defined $direct_parents{$tagtype}{$tagid}) {
+			
+			$taxonomy_ref->{$tagid}{parents} = [];
+			foreach my $parentid (sort keys %{$direct_parents{$tagtype}{$tagid}}) {
+				if ((not defined $fields_ref) or (defined $fields_ref->{parents})) {
+					exists $taxonomy_ref->{$tagid}{parents} or $taxonomy_ref->{$tagid}{parents} = [];
+					push @{ $taxonomy_ref->{$tagid}{parents} }, $parentid;
+				}
+				# Include parents if the tag is one of the initially requested tags
+				# so that we don't also add parents of parents
+				if (($options_ref->{include_parents}) and ($requested_tags{$tagid})) {
+					push @tags, $parentid;
+				}
+			}
+		}
+
+		if (defined $direct_children{$tagtype}{$tagid}) {
+			
+			foreach my $childid (sort keys %{$direct_children{$tagtype}{$tagid}}) {
+				if ((not defined $fields_ref) or (defined $fields_ref->{children})) {
+					exists $taxonomy_ref->{$tagid}{children} or $taxonomy_ref->{$tagid}{children} = [];
+					push @{$taxonomy_ref->{$tagid}{children}}, $childid;
+				}
+				if (($options_ref->{include_children}) and ($requested_tags{$tagid})) {
```

Comment:
I think you could repeat comments from above here.

Reply:
Agreed